discover/transport.go

```diff
@@ -0,0 +1,318 @@
+package discover
+
+import (
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"github.com/fluffle/goirc/client"
+	"github.com/roasbeef/btcd/btcec"
+	"github.com/roasbeef/btcd/wire"
+	"strconv"
+	"strings"
+	"time"
+)
+
+const (
+	// RPL_MOTD is an IRC WHOIS response. Unreal IRCd, which is the IRC
+	// daemon that runs Freenode, introduced it to denote the WHOIS Host,
+	// which is where we get our own IP from. It's not standardized, but
+	// useful for us to discover our external IP.
+	// See: https://www.alien.net.au/irc/irc2numerics.html#378
+	RPL_MOTD = "378"
+)
+
+// Transport interface is used to create an abstraction layer over
+// system-dependable representation of discovery service messages (IRC). It
+// gives the ability to discovery service operate only with final representation
+// of messages like ChanAnnouncement, NodeAnnouncement thereby decouple
+// this systems and give them ability to be easily testable.
+type Transport interface {
+	Connect() (string, error)
+
+	// Joined returns the channel which will be triggered when some new
+	// peer joined.
+	Joined() chan bool
```

Comment:
Re-announcing our entire state each time we learn of a new participant seems like it would be extremely inefficient. Instead I was under the impression that everything is to be based on randomized timers with batched updates.